### ISSUE TITLE
[pkg] fetch dependencies without building the project

### ISSUE THREAD

**Issue by maiste** [ocaml/dune#10949]
## Context

In the current behaviour, unless you have the exact target, it is impossible to download the dependencies without building the project. A nice feature would be a command that fetch the project dependencies in the same way you would with `dune build`. This is important in the context of Dockerfile, where you can save some computation time by caching this stage. 

## Solution

We can implement a `dune pkg fetch` command that triggers the fetch rule for the entire project.

**Comment by rgrinberg** [comment]
What about just introducing an alias `@pkg-fetch`? This is the more standard mechanism for collecting targets in dune.

**Comment by maiste** [comment]
Yes, definitely! It makes more sense. Another proposition was `dune build @deps`.

The original question was if there was a way to mimic the behaviour of `opam install --deps-only --with-test .` in the context of Dune Package Management. It means that it should fetch the target and build them too. However, I don't know if we have the possibility to execute the test part?

**Comment by leostera** [comment]
To clarify, we have 2 use-cases here.

1. Writing a Dockerfile where we can stage building dependencies separate from building the entire project, and cache that based off the dune-project (or eventually the dune.lock):

```dockerfile
FROM ubuntu

RUN curl https://get.ocaml.org | bash 

# Adds project + deps
ADD dune-project .
RUN dune pkg lock

# Fetch and build dependencies only if the above changes
RUN dune build @deps 

# Adds rest of the sources
ADD . .
RUN dune build
```

2. A single command to fetch all external / downloadable resources for a project, to let you work without connectivity:

```sh
$ git clone github.com/my/project project
$ cd project
$ dune fetch
$ # now i can go offline
$ dune build 
```

**Comment by rgrinberg** [comment]
> The original question was if there was a way to mimic the behaviour of `opam install --deps-only --with-test .` in the context of Dune Package Management. It means that it should fetch the target and build them too. However, I don't know if we have the possibility to execute the test part?

We already build the test deps. So `dune runtest` should work.

**Comment by maiste** [comment]
I have work on part of this in #11046 where it introduces an alias to install and build packages. However, as stated here, we would like to have another alias to just fetch the dependencies but without building them. It is currently unclear to me what would be the proper way to address this second part. Indeed, the call to [Fetch.fetch](https://github.com/ocaml/dune/blob/b61e9d612642dbd1fd58d25c90126d607e69bd43/src/dune_pkg/source.ml#L39) in source.ml seems to also build the dependency. 

@rgrinberg would you have a suggestion on where I could introduce the alias? 

Does it mean that I have to split the function in two parts so we prevent the building?

**Comment by rgrinberg** [comment]
Is this issue fixed now?

**Comment by maiste** [comment]
Yes, it was fixed by #11046.
